src/wallet/asyncrpcoperation_saplingmigration.cpp

```diff
@@ -0,0 +1,196 @@
+#include "assert.h"
+#include "boost/variant/static_visitor.hpp"
+#include "asyncrpcoperation_saplingmigration.h"
+#include "init.h"
+#include "rpc/protocol.h"
+#include "random.h"
+#include "sync.h"
+#include "tinyformat.h"
+#include "transaction_builder.h"
+#include "util.h"
+#include "wallet.h"
+
+const CAmount FEE = 10000;
+
+AsyncRPCOperation_saplingmigration::AsyncRPCOperation_saplingmigration(int targetHeight) : targetHeight_(targetHeight) {}
+
+AsyncRPCOperation_saplingmigration::~AsyncRPCOperation_saplingmigration() {}
+
+void AsyncRPCOperation_saplingmigration::main() {
+    if (isCancelled())
+        return;
+
+    set_state(OperationStatus::EXECUTING);
+    start_execution_clock();
+
+    bool success = false;
+
+    try {
+        success = main_impl();
+    } catch (const UniValue& objError) {
+        int code = find_value(objError, "code").get_int();
+        std::string message = find_value(objError, "message").get_str();
+        set_error_code(code);
+        set_error_message(message);
+    } catch (const runtime_error& e) {
+        set_error_code(-1);
+        set_error_message("runtime error: " + string(e.what()));
+    } catch (const logic_error& e) {
+        set_error_code(-1);
+        set_error_message("logic error: " + string(e.what()));
+    } catch (const exception& e) {
+        set_error_code(-1);
+        set_error_message("general exception: " + string(e.what()));
+    } catch (...) {
+        set_error_code(-2);
+        set_error_message("unknown error");
+    }
+
+    stop_execution_clock();
+
+    if (success) {
+        set_state(OperationStatus::SUCCESS);
+    } else {
+        set_state(OperationStatus::FAILED);
+    }
+
+    std::string s = strprintf("%s: Sprout->Sapling transactions sent. (status=%s", getId(), getStateAsString());
+    if (success) {
+        s += strprintf(", success)\n");
+    } else {
+        s += strprintf(", error=%s)\n", getErrorMessage());
+    }
+
+    LogPrintf("%s", s);
+}
+
+bool AsyncRPCOperation_saplingmigration::main_impl() {
+    std::vector<CSproutNotePlaintextEntry> sproutEntries;
+    std::vector<SaplingNoteEntry> saplingEntries;
+    {
+        LOCK2(cs_main, pwalletMain->cs_wallet);
+        // Consider, should notes be sorted?
+        pwalletMain->GetFilteredNotes(sproutEntries, saplingEntries, "", 0);
+    }
+    if (sproutEntries.empty()) { // Consider, should the migration remain enabled?
```

Comment:
If not then that should be added to ZIP 308.

Leaving it enabled might be more convenient: the user could receive further funds to Sprout addresses, and probably also wants those migrated.

On the other hand, I'm concerned that migration is left enabled then eventually there won't be a large enough anonymity set of other migration transactions to hide in. I'm not sure what we can do about that.

Reply:
This could be covered by, "if the remaining amount to be migrated is less than 0.01 ZEC, end the migration," but we could more explicit. I agree that some users may want to leave this enabled, so this could be revisited, but eventually sending funds to a Sprout address will be disabled and this will become a moot point.